base/thread_pool.h

```diff
@@ -110,6 +110,8 @@ class thread_pool {
   // Waits for `condition` to become true. While waiting, executes tasks on the queue.
   // The condition is executed atomically.
   virtual void wait_for(predicate_ref condition) = 0;
+  virtual void wait_for_flag(predicate_ref condition, std::atomic<bool>& flag) = 0;
```

Comment:
I think this is still waiting for the predicate, just with this extra flag to wake it up. Should this just be an overload?